apps/studio/src/features/editing-experience/components/ComplexEditorStateDrawer.tsx

```diff
@@ -104,6 +109,87 @@ export default function ComplexEditorStateDrawer(): JSX.Element {
     setPreviewPageState(newPageState)
   }
 
+  const handleSave = async () => {
+    let newPageState = previewPageState
+
+    if (modifiedAssets.length > 0) {
+      const updatedBlocks = Array.from(previewPageState.content)
+      const newBlock = _.cloneDeep(updatedBlocks[currActiveIdx])
+
+      if (!newBlock) {
+        return
+      }
+
+      // Upload all new/modified images/files
+      const assetsToUpload = modifiedAssets.filter((asset) => !!asset.file)
+
+      const isUploadingSuccessful = await Promise.allSettled(
+        assetsToUpload.map(({ path, file }) => {
+          if (!file) {
+            return
+          }
+
+          return uploadAsset({ file }).then((res) => {
+            _.set(newBlock, path, res.path)
+            return path
+          })
+        }),
+      ).then((results) => {
+        // Keep only failed uploads inside modifiedAssets so on subsequent
+        // save attempts, we retry uploading just the failed assets
+        const newModifiedAssets = modifiedAssets.filter(({ path }) => {
+          return !results.some(
+            (result) => result.status === "fulfilled" && result.value === path,
+          )
+        })
+
+        if (newModifiedAssets.length > 0) {
+          toast({
+            title: "Error uploading assets",
+            description:
+              "An error occurred while uploading some assets. Please try again later.",
```

Comment:
For clarity: Should we tell the user something like 2/10 assets failed to upload. Please try again.

cc @sehyunidaaa

Reply:
Hmm opted not to include this information because there is no further feedback to the user on which assets are the ones that are failing, hence keeping it generic.

Reply:
when does this error occur? where would a user upload multiple assets?

Reply:
nits regardless of the context (i assumed we only support 1 image at a time now): 
- would avoid the word "assets" for user-facing copies! files or images if can be specific. if not, `An error occurred while uploading files/images. Try again later.` would also do
- if Harish's recommendation isn't doable, would be great if there's some (visual) indication on the studio which ones failed vs. which ones succeeded! that way the user can recover from the error by retrying the ones that failed

Reply:
Yup @sehyunidaaa in general we will only have 1 image at a time, but this also applies when the particular component can have multiple images (e.g. multiple infocards with images). I can show the number of failed images but it might not be easy to show exactly which component has errors.

Adjusted in 2fd161656ed3c823494b80645dce02cd1fa8af92!